org.eclipse.scanning.api/src/org/eclipse/scanning/api/points/AbstractGenerator.java

```diff
@@ -117,6 +124,83 @@ public final int size() throws GeneratorException {
 		validateModel();
 		return sizeOfValidModel();
 	}
+	
+	public int getRank() throws GeneratorException {
+		return getShape().length;
+	}
+	
+	public int[] getShape() throws GeneratorException {
+		if (shape == null) {
+			shape = calculateShape();
+		}
+		
+		return shape;
+	}
+	
+	/**
+	 * Calculates the shape of the scan. This method is called when
+	 * {@link #iteratorFromValidModel()} does not return a {@link ScanPointIterator}.
+	 * Subclasses should override if a more efficient way of calculating the
+	 * scan shape can be provided.
+	 *  
+	 * @param iterator
+	 * @return
+	 * @throws GeneratorException
+	 */
+	protected int[] calculateShape() throws GeneratorException {
+		Iterator<IPosition> iterator = iteratorFromValidModel();
+		if (iterator instanceof ScanPointIterator) {
+			return ((ScanPointIterator) iterator).getShape();
+		}
+		
+		if (iterator.hasNext()) {
+			return new int[0];
+		}
+		
+		IPosition first = iterator.next();
+		final int scanRank = first.getScanRank();
+		if (scanRank == 0) {
+			return new int[0];
+		}
+		
+		int pointNum = 1;
+
+		IPosition last = first;
+		int lastInnerIndex = last.getIndex(scanRank - 1);
+		int maxInnerIndex = -1;
+		long lastTime = System.currentTimeMillis();
+		while (iterator.hasNext()) {
+			last = iterator.next(); // Could be large...
+			pointNum++;
+			if (maxInnerIndex == -1) {
+				// check for the max index of the inner most dimension - we assume that when it
+				// first decreases the previous value was the max. This special treatment
+				int innerIndex = last.getIndex(scanRank - 1);
+				if (innerIndex <= lastInnerIndex) {
+					maxInnerIndex = lastInnerIndex;
+				}
+			}
+			
+			if (pointNum % 10000 == 0) {
+				long newTime = System.currentTimeMillis();
+				System.err.println("Point number " + pointNum++ + ", took " + (newTime - lastTime) + "ms");
+			}
+		}
+		
+		// special case for scans of rank 0 - i.e. acquire scans
+		if (scanRank == 0) return new int[0];
+		
+		// the shape is created from the indices for each dimension for the final scan point
+		int[] shape = new int[scanRank];
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Rename "shape" which hides the field declared at line 42. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AHiddenFieldCheck)